### ISSUE TITLE
[Master Branch OC4] Please re-work extension namespaces and other extension issues

### ISSUE THREAD

**Issue by mhcwebdesign** [opencart/opencart#9591]
The extension name spaces are quite weird at the moment. For example, the **extension/opencart/admin/controller/module/account.php** currently uses this namespace:

`namespace Opencart\Admin\Controller\Extension\Opencart\Module;`

The namespaces should really fall in line with the underlying folder structure, as is common with many other packages like Symfony, PHPWord, or with Java packages. I suggest the following namespace convention for OC4 extensions:

OC4 extension namespaces should always start with Opencart\Extension\<extension-name>\<application>\

For example, above module would use this:

`namespace Opencart\Extension\Opencart\Admin\Controller\Module;`

If a 3rd party extension were to introduce a new extension names 'foo' with a module named 'bar' (using file `upload/extension/foo/admin/controller/module/bar.php` in its ocmod.zip archive) then its namespace should be:

`namespace Opencart\Extension\Foo\Admin\Controller\Module`

**Comment by mhcwebdesign** [comment]
Not even the opencart.ocmod.zip which contains various former core OC extensions can be properly installed or uninstalled without causing numerous errors, too many to be listed here. When I have some time, I'll take a closer look at it this weekend.

**Comment by chongshengdz** [comment]
You are very very wrong.
If a 3rd party extension were to introduce a new extension package named 'foo' with a module named 'bar' (using file upload/extension/foo/admin/controller/module/bar.php in its ocmod.zip archive) then its namespace should be:

it should be 

namespace Opencart\Admin\Controller\Extension\Foo\Module;
class Bar extends \Opencart\System\Engine\Controller {


and it should be **foo.ocmod.zip**

**Comment by chongshengdz** [comment]
just tested the master version and found it still not stable, cannot save data in the Admin.
below is a sample extension, you can download and test it.
https://www.ectransistors.com/groupfolder.ocmod.zip

**Comment by stalker780** [comment]
@chongshengdz 
Don't waste your time. Latest stable OC4 was before admin ajax commits.
I am working with a copy from 26/March.
At least it could be called beta then.

Now it looks like raw alpha again, with lots of bugs, typos and unfinished functionality in admin area.

**Comment by chongshengdz** [comment]
> @chongshengdz
> Don't waste your time. Latest stable OC4 was before admin ajax commits.
> I am working with a copy from 26/March.
> At least it could be called beta then.
> 
> Now it looks like raw alpha again, with lots of bugs, typos and unfinished functionality in admin area.

thanks for the information, will try again later.

**Comment by mhcwebdesign** [comment]
> it should be
> 
> namespace Opencart\Admin\Controller\Extension\Foo\Module;
> class Bar extends \Opencart\System\Engine\Controller {

You are wrong.  The namespace needs to reflect the underlying folder structure, making it easier for autoloaders to find the class file.

**Comment by mhcwebdesign** [comment]
Or, to reflect Daniel's latest namespace changes in his **opencart.ocmod.zip** extension files, the folder structure for an extension named 'foo' should be:

<opencart-folder>/admin/controller/extension/foo/*
<opencart-folder>/admin/language/*/extension/foo/*
<opencart-folder>/admin/model/extension/foo/*
<opencart-folder>/admin/view/template/extension/foo/*
<opencart-folder>/catalog/controller/extension/foo/*
<opencart-folder>/catalog/language/*/extension/foo/*
<opencart-folder>/catalog/model/extension/foo/*
<opencart-folder>/catalog/view/template/extension/foo/*

and we won't need folder
<opencart-folder>/extension/*
at all.

**Comment by mhcwebdesign** [comment]
Not really, the extension namespaces still don't reflect the underlying folder structure!

**Comment by danielkerr** [comment]
ytes for extensions it changes
